File: Geek Assistant (Android AIO)/Modules/Android/Action/AutoDetect.vb
Imports Managed

Module AutoDetect

    Private ErrorInfo As (lvl As Integer, msg As String)
    Public Sub Run(Optional Silent As Boolean = False)

        ' (ROOT) 
        ' Remove screen lock: "adb shell su -c rm /data/system/*.key" && "adb shell su -c rm /data/system/locksettings*"
        ' Hot reboot: "adb shell su -c busybox killall system_server"
        ' Clear dulvik cache: "adb shell su -c rm -R /data/dalvik-cache"
        ' Factory Reset: "adb shell su -c recovery --wipe_data"

        Working = True
        ErrorCodeTrack("AD-00") ' adb Auto - Begin
        If Not Silent Then LogEvent("Auto Detect", 2)
        Main.ProgressBar.Value = 0
        Try

            If Not Silent Then GA_SetProgressText.Run("Clearing previous device information.", -1)
            ErrorInfo = (10, "We had trouble while clearing previous device information.")
            ResetDeviceInfo()
            Main.ProgressBar.Value = 2

            If Not Silent Then GA_SetProgressText.Run("Preparing the environment... Please be patient.", -1)
            ErrorInfo = (10, "Things didn't go as planned while preparing the environment.")
            madbStart()
            Main.ProgressBar.Value = 5

            If Not Silent Then GA_SetProgressText.Run("Counting the connected devices...", -1)
            ErrorInfo = (10, "Math...oh no we couldn't count the devices.")
            If madb_GetDeviceCount() = 0 Then
                Main.DeviceState_Label.Text = "Disconnected"
                ErrorInfo = (0, $"We haven't found any device.{vbCrLf}{My.Resources.TroubleshootConnection}")
                ErrorCodeTrack("AD-D0") ' adb Auto - Device 0 (0 devices connected)
                Throw New Exception
            ElseIf madb_GetDeviceCount() > 1 Then
                Main.DeviceState_Label.Text = "Multiple"
                ErrorInfo = (0, $"Oh there are several devices.{vbCrLf}Would you mind keeping 1 and disconnecting the rest please?")
                ErrorCodeTrack("AD-DX") ' adb Auto - Device X-number (More than 1 connected)
                Throw New Exception
            End If
            'DeviceInfoSave(Nothing, "Connected", Nothing, -1, 0, 0, 0)

            Main.ProgressBar.Value = 7
            If Not Silent Then GA_SetProgressText.Run("Communicating with your device...", -1)
            ErrorInfo = (1, "We have trouble reading your device.")

            Dim DeviceState_String As String = "Device is "

            Select Case madb_GetDeviceState()
                Case 5 'unknown
                    Main.DeviceState_Label.Text = "Unknown"
                    DeviceState_String &= $"in an unknown state...{vbCrLf}{My.Resources.TroubleshootConnection}"
                    ErrorInfo = (-1, DeviceState_String)
                    ErrorCodeTrack("AD-DU") ' adb Auto - Device 0 (No devices connected)
                    Throw New Exception

                Case 2 'offline
                    Main.DeviceState_Label.Text = "Offline"
                    DeviceState_String &= $"offline. {vbCrLf}{My.Resources.TroubleshootConnection}"
                    ErrorInfo = (0, DeviceState_String)
                    ErrorCodeTrack("AD-DO") ' adb Auto - Device Offline (PC not allowed to debug device)
                    Throw New Exception
                Case 0 'recovery
                    Main.DeviceState_Label.Text = "Recovery mode"
                    DeviceState_String &= $"in recovery mode.{vbCrLf}Please enter adb mode and try again." 'Please enter adb mode or reboot to system and try again."
                    ErrorInfo = (0, DeviceState_String)
                    ErrorCodeTrack("AD-DR") ' adb Auto - Device Recovery
                    Throw New Exception
                Case 4 'download
                    Main.DeviceState_Label.Text = "Download mode"
                    DeviceState_String &= $"in download mode.{vbCrLf}Please enter adb mode and try again."
                    ErrorInfo = (0, DeviceState_String)
                    ErrorCodeTrack("AD-DD") ' adb Auto - Device Download
                    Throw New Exception

            ' ^^   All the above will jump to > Catch e As Exception   ^^
            ' vv            All the below will continue code           vv

                Case 1 'bootloader 
                    DeviceState_String &= $"in fastboot mode."
                    Main.DeviceState_Label.Text = "Fastboot mode"
                    ErrorCodeTrack("AD-DF") ' adb Auto - Device Fastboot
                    Main.ProgressBar.Value = 10
                    If Not Silent Then
                        Dim DeviceInFastboot_ContinueAsk = GA_infoAsk.Run($"{DeviceState_String}",
                                                                      $"We cannot read much in this mode.{vbCrLf}Do you want to continue detection in fastboot mode?",
                                                                       "Continue", "Cancel")
                        If DeviceInFastboot_ContinueAsk Then


                            ErrorInfo.lvl = 1
                            DoMsg(1, $"Oh no this is currently unavailable.{vbCrLf}{My.Resources.FeatureUnavailable }", 2)
                            'later maybe will be implemented 
                            '''''''''''''''''''''''''''''''''

                        Else
                            GA_SetProgressText.Run("Detection cancelled by user.", 0)
                        End If
                    End If

                Case 3 'online
                    ErrorCodeTrack("AD-D1") ' adb Auto - Device 1 (1 Device connected in adb mode)
                    Main.ProgressBar.Value = 10
                    ErrorInfo.lvl = 10

                    Dim dev As Adb.Device = madb_GetListOfDevice(0) 'Set dev as first device

                    DeviceState_String &= $"in adb mode."
                    Main.DeviceState_Label.Text = "Connected (ADB)"
                    Main.ProgressBar.Value = 11


                    If Not Silent Then GA_SetProgressText.Run("Retrieving device manufacturer, model, and codename...", -1)
                    Dim Manufacturer_CommandResultReceiver As New Adb.CommandResultReceiver
                    Main.ProgressBar.Value = 13
                    dev.ExecuteShellCommand($"getprop ro.product.manufacturer", Manufacturer_CommandResultReceiver)
                    Main.ProgressBar.Value = 15
                    My.Settings.DeviceManufacturer = FixManufacturerString(Manufacturer_CommandResultReceiver.Result)
                    My.Settings.Save()
                    Main.Manufacturer_ComboBox.Text = My.Settings.DeviceManufacturer
                    If Not Silent Then LogAppendText($" > {My.Settings.DeviceManufacturer} {dev.Model} ({dev.Product})", 1)
                    Main.ProgressBar.Value = 17

                    If Not Silent Then GA_SetProgressText.Run("Retrieving device serial#...", -1)
                    My.Settings.DeviceSerial = dev.SerialNumber
                    My.Settings.Save()
                    If Not Silent Then LogAppendText($" | Serial: {My.Settings.DeviceSerial}", 1)
                    Main.ProgressBar.Value = 20

                    If Not Silent Then GA_SetProgressText.Run("Retrieving root state...", -1)
                    My.Settings.DeviceRooted = dev.CanSU
                    My.Settings.Save()
                    Main.Rooted_Checkbox.Checked = My.Settings.DeviceRooted
                    If Not Silent Then LogAppendText($" | Rooted: {txt_ConvertBooleanYesNo(My.Settings.DeviceRooted)}", 1)
                    Main.ProgressBar.Value = 25

                    If Not Silent Then GA_SetProgressText.Run("Retrieving bootloader unlock support state...", -1)
                    Dim BLunlockable_CommandResultReceiver As New Adb.CommandResultReceiver
                    Main.ProgressBar.Value = 26
                    dev.ExecuteShellCommand($"getprop ro.oem_unlock_supported", BLunlockable_CommandResultReceiver)
                    My.Settings.DeviceBootloaderUnlockSupported = BLunlockable_CommandResultReceiver.Result
                    My.Settings.Save()
                    Main.BootloaderUnlockable_CheckBox.Checked = My.Settings.DeviceBootloaderUnlockSupported
                    Main.ProgressBar.Value = 27
                    If Not Silent Then LogAppendText($" | Bootloader Unlock allowed: {txt_ConvertBooleanYesNo(My.Settings.DeviceBootloaderUnlockSupported)}", 1)
                    Main.ProgressBar.Value = 30

                    If Not Silent Then GA_SetProgressText.Run("Retrieving Android API level...", -1)
                    Dim API_CommandResultReceiver As New Adb.CommandResultReceiver
                    Main.ProgressBar.Value = 31
                    dev.ExecuteShellCommand($"getprop {Adb.Device.PROP_BUILD_API_LEVEL}", API_CommandResultReceiver)
                    Main.ProgressBar.Value = 32
                    My.Settings.DeviceAPILevel = Convert.ToInt32(API_CommandResultReceiver.Result)
                    My.Settings.Save()
                    Main.AndroidVersion_ComboBox.Text = ConvertAPILevelToAVer(My.Settings.DeviceAPILevel)(1)
                    If Not Silent Then GA_SetProgressText.Run("Converting API level to Android name...", -1)
                    Main.ProgressBar.Value = 33
                    If Not Silent Then LogAppendText($" | Android Version: {ConvertAPILevelToAVer(My.Settings.DeviceAPILevel)(0)} (API: {My.Settings.DeviceAPILevel})", 1)
                    Main.ProgressBar.Value = 35

                    If Not Silent Then GA_SetProgressText.Run("Retrieving battery level...", -1)
                    Main.ProgressBar.Value = 36
                    Dim batteryString As String
                    If dev.GetBatteryInfo.Present Then
                        My.Settings.DeviceBatteryLevel = dev.GetBatteryInfo.Level
                        batteryString = $"{My.Settings.DeviceBatteryLevel}%"
                    Else
                        My.Settings.DeviceBatteryLevel = -1 'no level. Not present
                        If Not Silent Then GA_SetProgressText.Run("Battery not present!", -1)
                        batteryString = "❌"
                    End If
                    My.Settings.Save()
                    Main.ProgressBar.Value = 38
                    If Not Silent Then LogAppendText($" | Battery: {batteryString}", 1)
                    Main.ProgressBar.Value = 40

                    If Not Silent Then GA_SetProgressText.Run(DeviceState_String, -1) 'This is after retrieving info to stay written in Progress Text



            End Select

            Main.ProgressBar.Value = 100
        Catch ex As Exception
            GA_PleaseWait.Run(False) 'Close before error dialog
            If Not Silent Then
                DoMsg(ErrorInfo.lvl, ErrorInfo.msg, 2, ex.ToString)
            Else
                Main.DoNeutral()
            End If
        End Try
        My.Settings.DeviceState = Main.DeviceState_Label.Text
        GA_PleaseWait.Run(False) 'Close if Try was successful
        Working = False
    End Sub

End Module

File: Geek Assistant (Android AIO)/Modules/Android/Action/AutoDetect.cs
using GeekAssistant.Forms;
using Managed.Adb;
using System;
using System.Windows.Forms;

// (ROOT)
// Remove screen lock: $"{workCode}b shell su -c rm /data/system/*.key" && $"{workCode}b shell su -c rm /data/system/locksettings*"
// Hot reboot: $"{workCode}b shell su -c busybox killall system_server"
// Clear dulvik cache: $"{workCode}b shell su -c rm -R /data/dalvik-cache"
// Factory Reset: $"{workCode}b shell su -c recovery --wipe_data"
// Root check: << "adb shell su" example >> "su: not found" 
// Root check: << set uid= for /f "delims=" %%a in ('adb -s devicename shell "su 0 id -u 2>/dev/null"') do set uid=%%a && echo %uid%
//             >> 0 or 1

internal static partial class AutoDetect {
    private static string workCode => $"{workCode}";
    private static string currentTitle => "Auto Detect";

    private static bool RunningInBetween = false;
    public static void Run(bool Silent = false) {
        //Refresh current Home instance
        Home Home = null;
        foreach (Form home in Application.OpenForms)
            if (home.GetType() == typeof(Home))
                Home = (Home)home;
        RunningInBetween = c.Working; //true if auto detecting while running another process
        c.Working = true;

        if (!RunningInBetween) inf.currentTitle = currentTitle;
        inf.detail.workCode = $"{(RunningInBetween ? $"{inf.detail.workCode}-" : "")}AD-00"; // (RunningInBetween? {workCode} -) Auto Detect - Begin
        if (!Silent)
            GA_Log.LogEvent(currentTitle, 2);

        Home.bar.Value = 0;
        try {
            if (!Silent)
                GA_SetProgressText.Run("Clearing previous device information.", -1);

            inf.detail = ($"{workCode}-CD", inf.lvls.FatalError, inf.currentTitle, "We had trouble while clearing previous device information.", null); // Auto Detect - Clear Device
            GA_adb.ResetDeviceInfo();

            Home.bar.Value = 2;
            if (!Silent)
                GA_SetProgressText.Run("c.Preparing the environment... Please be patient.", -1);

            inf.detail = ($"{workCode}-PE", inf.lvls.FatalError, inf.currentTitle, "Things didn't go as planned while preparing the environment.", null); // Auto Detect - Prepare Environment
            madb.madbBridge();

            Home.bar.Value = 5;
            if (!Silent)
                GA_SetProgressText.Run("Counting the connected devices...", -1);

            inf.detail = ($"{workCode}-Dc", inf.lvls.FatalError, inf.currentTitle, "Math...oh no we couldn't count the devices.", null); // Auto Detect - Device count X
            switch (madb.GetDeviceCount()) {
                case 0:
                    Home.DeviceState_Label.Text = "Disconnected";
                    inf.detail = ($"{workCode}-D0", inf.lvls.Warn, inf.currentTitle, $"We haven't found any device.\n{prop.strings.TroubleshootConnection}", null); // Auto Detect - Device 0 (0 devices connected)
                    throw new Exception();
                case > 1:
                    Home.DeviceState_Label.Text = "Multiple";
                    inf.detail = ($"{workCode}-DX", inf.lvls.Warn, inf.currentTitle, $"Oh there are several devices.\nWould you mind keeping 1 and disconnecting the rest please?", null); // Auto Detect - Device X-number (More than 1 connected)
                    throw new Exception();
            }

            Home.bar.Value = 7;
            if (!Silent)
                GA_SetProgressText.Run("Communicating with your device...", -1);

            inf.detail = ($"{workCode}-Ds", inf.lvls.Warn, inf.currentTitle, "We have trouble reading your device.", null); // Auto Detect - Device count (failed to count devices)
            string DeviceState_String = "Device is ";
            switch (madb.GetDeviceState()) {
                case 5: // unknown 
                    Home.DeviceState_Label.Text = "Unknown";
                    DeviceState_String += $"in an unknown state...\n{prop.strings.TroubleshootConnection}";
                    inf.detail = ($"{workCode}-DU", inf.lvls.Information, inf.currentTitle, DeviceState_String, null); // Auto Detect - Device 0 (No devices connected)
                    throw new Exception();
                case 2: // offline 
                    Home.DeviceState_Label.Text = "Offline";
                    DeviceState_String += $"offline. \n{prop.strings.TroubleshootConnection}";
                    inf.detail = ($"{workCode}-DO", inf.lvls.Warn, inf.currentTitle, DeviceState_String, null); // Auto Detect - Device Offline (PC not allowed to debug device)
                    throw new Exception();
                case 0: // recovery 
                    Home.DeviceState_Label.Text = "Recovery mode";
                    DeviceState_String += $"in recovery mode.\nPlease enter adb mode and try again."; // Please enter adb mode or reboot to system and try again."
                    inf.detail = ($"{workCode}-DR", inf.lvls.Warn, inf.currentTitle, DeviceState_String, null); // Auto Detect - Device Recovery
                    throw new Exception();
                case 4: // download 
                    Home.DeviceState_Label.Text = "Download mode";
                    DeviceState_String += $"in download mode.\nPlease enter adb mode and try again.";
                    inf.detail = ($"{workCode}-DD", inf.lvls.Warn, inf.currentTitle, DeviceState_String, null); // Auto Detect - Device Download
                    throw new Exception();

                // ^^   All the above will jump to >> catch (Exception ex) { >>  ^^
                // vv              All the below will continue code              vv

                case 1: // bootloader 
                    DeviceState_String += $"in fastboot mode.";
                    Home.DeviceState_Label.Text = "Fastboot mode";
                    inf.detail.workCode = $"{workCode}-DF"; // Auto Detect - Device Fastboot
                    Home.bar.Value = 10;
                    if (!Silent) {
                        var DeviceInFastboot_ContinueAsk =
                            inf.Run(inf.lvls.Question, $"{DeviceState_String}",
                                      $"We cannot read much in this mode.\nDo you want to continue detection in fastboot mode?",
                                    ("Continue", "Close"));
                        if (DeviceInFastboot_ContinueAsk) {
                            inf.detail.lvl = inf.lvls.Error;
                            inf.Run(inf.lvls.Error, inf.currentTitle,
                                   $"Oh no this is currently unavailable.\n{prop.strings.FeatureUnavailable}");
                        }
                        // later maybe will be implemented
                        // ''''''''''''''''''''''''''''''''
                        else {
                            GA_SetProgressText.Run("Detection cancelled by user.", 0);
                        }
                    }
                    break;

                case 3: // online 
                    Home.bar.Value = 10;
                    inf.detail = ($"{workCode}-DlX", inf.lvls.Error, inf.currentTitle, "Sorry we can't see your device anymore...", null);// Auto Detect - Device list X
                    Device dev = madb.GetListOfDevice()[0]; // Set dev as first device
                    DeviceState_String += $"in adb mode.";
                    Home.DeviceState_Label.Text = "Connected (ADB)";
                    Home.bar.Value = 11;

                    inf.detail = ($"{workCode}-D-m", inf.lvls.Error, inf.currentTitle, "Failed to check your device manufacturer information.", null); // Auto Detect - Device - manufacturer
                    if (!Silent)
                        GA_SetProgressText.Run("Fetching device manufacturer, model, and codename...", -1);

                    Home.bar.Value = 13;

                    Home.bar.Value = 15;
                    c.S.DeviceManufacturer =
                        GA_adb.FixManufacturerString(
                            cmd.madbShell(dev, "getprop ro.product.manufacturer"));
                    c.S.Save();
                    Home.Manufacturer_ComboBox.Text = c.S.DeviceManufacturer;

                    inf.detail = ($"{workCode}-D-mc", inf.lvls.Error, inf.currentTitle, "Failed to check your device model or codename.", null);// Auto Detect - Device - model codename
                    if (!Silent)
                        GA_Log.LogAppendText($" ❱ {c.S.DeviceManufacturer} {dev.Model} ({dev.Product})", 1);

                    Home.bar.Value = 17;

                    inf.detail = ($"{workCode}-D-s", inf.lvls.Error, inf.currentTitle, "Failed to check your device serial number.", null);// Auto Detect - Device - serial
                    if (!Silent)
                        GA_SetProgressText.Run("Fetching device serial#...", -1);

                    c.S.DeviceSerial = dev.SerialNumber;
                    c.S.Save();
                    if (!Silent)
                        GA_Log.LogAppendText($" | Serial: {c.S.DeviceSerial}", 1);

                    Home.bar.Value = 20;

                    inf.detail = ($"{workCode}-D-su", inf.lvls.Error, inf.currentTitle, "Failed to check your device root status.", null);// Auto Detect - Device - su
                    if (!Silent)
                        GA_SetProgressText.Run("Fetching root state...", -1);

                    c.S.DeviceRooted = dev.CanSU();
                    c.S.Save();
                    Home.Rooted_Checkbox.Checked = c.S.DeviceRooted;
                    if (!Silent)
                        GA_Log.LogAppendText($" | Rooted: {convert.Bool.ToYesNo(c.S.DeviceRooted)}", 1);

                    Home.bar.Value = 23;

                    inf.detail = ($"{workCode}-D-bb", inf.lvls.Error, inf.currentTitle, "Failed to check your device busybox availability.", null);// Auto Detect - Device - busybox
                    if (!Silent)
                        GA_SetProgressText.Run("Fetching busybox availability...", -1);

                    c.S.DeviceBusyBoxReady = dev.BusyBox.Available;
                    c.S.Save();
                    if (!Silent)
                        GA_Log.LogAppendText($" | Busybox available: {convert.Bool.ToYesNo(c.S.DeviceBusyBoxReady)}", 1);

                    Home.bar.Value = 25;

                    inf.detail = ($"{workCode}-D-blu", inf.lvls.Error, inf.currentTitle, "Failed to check your device bootloader unlock support.", null); // Auto Detect - Device - bootloader unlock
                    if (!Silent)
                        GA_SetProgressText.Run("Fetching bootloader unlock support state...", -1);

                    Home.bar.Value = 26;
                    c.S.DeviceBootloaderUnlockSupported =
                         convert.String.ToBool(cmd.madbShell(dev, "getprop ro.oem_unlock_supported"));
                    c.S.Save();
                    Home.BootloaderUnlockable_CheckBox.Checked = c.S.DeviceBootloaderUnlockSupported;
                    Home.bar.Value = 27;
                    if (!Silent)
                        GA_Log.LogAppendText($" | Bootloader unlock allowed: {convert.Bool.ToYesNo(c.S.DeviceBootloaderUnlockSupported)}", 1);

                    Home.bar.Value = 30;

                    inf.detail = ($"{workCode}-D-al", inf.lvls.Error, inf.currentTitle, "Failed to check your device API level.", null);// Auto Detect - Device - API level
                    if (!Silent)
                        GA_SetProgressText.Run("Fetching Android API level...", -1);

                    Home.bar.Value = 32;
                    c.S.DeviceAPILevel = Convert.ToInt32(cmd.madbShell(dev, $"getprop {Device.PROP_BUILD_API_LEVEL}"));
                    c.S.Save();
                    inf.detail = ($"{workCode}-D-atv", inf.lvls.Error, inf.currentTitle, "Failed to convert the API level to Android version.", null);// Auto Detect - Device - API level version
                    Home.AndroidVersion_ComboBox.Text = GA_adb.ConvertAPILevelToAVer(c.S.DeviceAPILevel)[1];
                    if (!Silent)
                        GA_SetProgressText.Run("Converting API level to Android name...", -1);

                    Home.bar.Value = 33;
                    if (!Silent)
                        GA_Log.LogAppendText($" | Android version: {GA_adb.ConvertAPILevelToAVer(c.S.DeviceAPILevel)[0]} (API: {c.S.DeviceAPILevel})", 1);

                    Home.bar.Value = 35;

                    inf.detail = ($"{workCode}-D-b", inf.lvls.Error, inf.currentTitle, "Failed to check your device battery level.", null); // Auto Detect - Device - battery
                    if (!Silent)
                        GA_SetProgressText.Run("Fetching battery level...", -1);

                    Home.bar.Value = 36;
                    string batteryString;
                    if (dev.GetBatteryInfo().Present) {
                        c.S.DeviceBatteryLevel = dev.GetBatteryInfo().Level;
                        batteryString = $"{c.S.DeviceBatteryLevel}%";
                    } else {
                        c.S.DeviceBatteryLevel = -1; // no level. Not present
                        if (!Silent)
                            GA_SetProgressText.Run("Battery not present!", -1);

                        batteryString = "❌";
                    }
                    c.S.Save();
                    Home.bar.Value = 38;
                    if (!Silent)
                        GA_Log.LogAppendText($" | Battery: {batteryString}", 1);

                    Home.bar.Value = 40;
                    if (!Silent)
                        GA_SetProgressText.Run(DeviceState_String, -1); // This is after retrieving info to stay written in Progress Text

                    break;
            }

            Home.bar.Value = 100;
        } catch (Exception ex) {
            if (!RunningInBetween) GA_Wait.Run(false); // Close before error dialog
            if (!Silent) {
                inf.detail.fullFatalError = ex.ToString();
                inf.Run();
            } else
                Home.DoNeutral();

        }

        c.S.DeviceState = Home.DeviceState_Label.Text;
        if (!RunningInBetween) GA_Wait.Run(false); // Close if Try was successful
        c.Working = RunningInBetween;
    }
}